gobblin-compaction/src/main/java/gobblin/compaction/parser/CompactionPathParser.java

```diff
@@ -0,0 +1,121 @@
+import gobblin.configuration.State;
+
+
+/**
+ * A parser which converts {@link FileSystemDataset} to {@link CompactionParserResult}
+ */
+public class CompactionPathParser extends CompactionParser<FileSystemDataset> {
+
+  public CompactionPathParser (State state) {
+    super (state);
+  }
+
+  /**
+   * A parsed result returned by {@link CompactionPathParser#parse(FileSystemDataset)}
+   */
+  public static class CompactionParserResult {
+    @Getter @Setter
+    private  String srcBaseDir;
+    @Getter @Setter
+    private  String dstBaseDir;
+    @Getter @Setter
+    private  String srcSubDir;
+    @Getter @Setter
+    private  String dstSubDir;
+
+    @Getter
+    private DateTime time;
+    @Getter
+    private String timeString;
+    @Getter
+    private String datasetName;
+  }
+
+  /**
+   * Parse a {@link FileSystemDataset} to some detailed parts like source base directory,
+   * source sub directory, destination based directory, destination sub directory, and time
+   * information.
+   */
+  public CompactionParserResult parse (FileSystemDataset dataset) {
+
+    CompactionParserResult result = new CompactionParserResult();
+    result.srcBaseDir = getSrcBaseDir (state);
+    result.srcSubDir  = getSrcSubDir  (state);
+    result.dstBaseDir = getDstBaseDir (state);
+    result.dstSubDir  = getDstSubDir  (state);
+
+    parseTimeAndDatasetName(dataset, result);
+
+    return result;
+  }
+
+  private void parseTimeAndDatasetName (FileSystemDataset dataset, CompactionParserResult rst) {
+    String commonBase = rst.getSrcBaseDir();
+    String fullPath = dataset.datasetURN();
+    int startPos = fullPath.indexOf(commonBase) + commonBase.length();
+    String relative = StringUtils.removeStart(fullPath.substring(startPos), "/");
+
+    int delimiterStart = StringUtils.indexOf(relative, rst.getSrcSubDir());
+    if (delimiterStart == -1) {
+      throw new StringIndexOutOfBoundsException();
+    }
+    int delimiterEnd = relative.indexOf("/", delimiterStart);
+    String datasetName = StringUtils.removeEnd(relative.substring(0, delimiterStart), "/");
+    String timeString = StringUtils.removeEnd(relative.substring(delimiterEnd + 1), "/");
+    rst.datasetName = datasetName;
+    rst.timeString = timeString;
+    rst.time = getTime (timeString);
+  }
+
+  private DateTime getTime (String timeString) {
+    DateTimeZone timeZone = DateTimeZone.forID(MRCompactor.DEFAULT_COMPACTION_TIMEZONE);
+    int splits = StringUtils.countMatches(timeString, "/");
+    String timePattern = "";
+    if (splits == 3) {
+      timePattern = "YYYY/MM/dd/HH";
+    } else if (splits == 2) {
+      timePattern = "YYYY/MM/dd";
+    }
+    DateTimeFormatter timeFormatter = DateTimeFormat.forPattern(timePattern).withZone(timeZone);
+    return timeFormatter.parseDateTime (timeString);
+  }
+
+  private String getSrcBaseDir(State state) {
+    Preconditions.checkArgument(state.contains(MRCompactor.COMPACTION_INPUT_DIR),
```

Comment:
I think adding Prrecondition here guarantees the correct configuration of state before we propagate the issue to the downstream method call (in case the next method swollen the error silently)